### ISSUE TITLE
Type narrowing on mutable attributes is wrong

### ISSUE THREAD

**Issue by kyle-query** [microsoft/pyright#7103]
**Describe the bug**

Pyright narrows types of mutable attributes when it should not.

**Code or Screenshots**

```python
from typing import Optional
from dataclasses import dataclass

@dataclass
class Foo:
    mutable: Optional[int] = None

    def mutate(self):
        self.mutable = 1

def test():
    f = Foo()

    if f.mutable is not None:
        return
    
    f.mutate()
    reveal_type(f.mutable)  # inferred type is None
```

Perhaps narrowing should still occur with `@dataclass(frozen=True)`, as I've found that very handy.

**VS Code extension or command-line**

Command line, pyright 1.1.348

**Comment by erictraut** [comment]
Pyright is working as designed here. This is not a bug or unintended behavior.

Pyright assumes for practical reasons that expressions whose types are narrowed within a local function will not be mutated "behind the back" of the function. That's generally a good assumption in well-written code. To assume otherwise would greatly limit the utility of type narrowing, so it's unlikely we would ever change this behavior. Most static type checkers in most languages (including mypy and the TypeScript compiler) make similar assumptions.
